File: app/src/main/java/com/nytimes/android/external/playbillingtester/di/ActivityModule.java
package com.nytimes.android.external.playbillingtester.di;

import android.app.Activity;
import android.app.AlertDialog;
import android.os.Environment;
import android.support.annotation.NonNull;

import com.google.common.io.Files;
import com.google.gson.Gson;
import com.nytimes.android.external.playbillingtester.BuildConfig;
import com.nytimes.android.external.playbillingtester.GithubApi;
import com.nytimes.android.external.playbillingtester.PermissionHandler;
import com.nytimes.android.external.playbillingtester.R;
import com.nytimes.android.external.playbillingtester.model.Config;
import com.nytimes.android.external.playbillingtester.model.ImmutableConfig;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileNotFoundException;

import javax.inject.Named;

import dagger.Module;
import dagger.Provides;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.google.common.base.Charsets.UTF_8;
import static com.nytimes.android.external.playbillingtester.APIOverrides.CONFIG_FILE;

@Module
public class ActivityModule {

    static final Logger LOGGER = LoggerFactory.getLogger(ActivityModule.class);

    private final Activity activity;

    public ActivityModule(@NonNull Activity activity) {
        this.activity = activity;
    }

    @Provides
    @ScopeActivity
    Activity provideActivity() {
        return activity;
    }


    @Provides
    @ScopeActivity
    Config provideConfig(Gson gson) {
        if (PermissionHandler.hasPermission(activity)) {
            try {
                return gson.fromJson(Files.newReader(new File(Environment.getExternalStorageDirectory().getPath(),
                        CONFIG_FILE), UTF_8), Config.class);
            } catch (FileNotFoundException exc) {
                LOGGER.error(activity.getString(R.string.config_not_found), exc);
            }
        } else {
            PermissionHandler.requestPermission(activity);
        }
        return ImmutableConfig.builder().build();
    }

    @Provides
    @ScopeActivity
    AlertDialog.Builder provideAlertDialogBuilder(Activity activity) {
        return new AlertDialog.Builder(activity);
    }

    @Provides
    @ScopeActivity
    GithubApi provideRetrofit(OkHttpClient client, @Named("gson_retrofit") Gson gson) {
        return new Retrofit.Builder()
                .client(client)
                .baseUrl(BuildConfig.GITHUB_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io()))
                .build()
                .create(GithubApi.class);
    }

}

File: app/src/main/java/com/nytimes/android/external/register/di/ActivityModule.kt
package com.nytimes.android.external.register.di

import android.app.Activity
import android.app.AlertDialog
import android.util.Log
import com.google.gson.Gson
import com.nytimes.android.external.register.APIOverrides.Companion.CONFIG_FILE
import com.nytimes.android.external.register.BuildConfig
import com.nytimes.android.external.register.GithubApi
import com.nytimes.android.external.register.R
import com.nytimes.android.external.register.di.RegisterApplicationModule.Companion.GSON_RETROFIT
import com.nytimes.android.external.register.model.Config
import dagger.Module
import dagger.Provides
import io.reactivex.schedulers.Schedulers
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.io.BufferedReader
import java.io.FileNotFoundException
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.Charset
import javax.inject.Named

@Module
class ActivityModule(private val activity: Activity) {

    @Provides
    @ScopeActivity
    internal fun provideActivity(): Activity {
        return activity
    }

    @Provides
    @ScopeActivity
    internal fun provideConfig(gson: Gson): Config? {
            try {
                return gson.fromJson(BufferedReader(InputStreamReader(readConfigFile(), Charset.forName("UTF-8"))), Config::class.java)
            } catch (exc: FileNotFoundException) {
                Log.e("ActivityModule", activity.getString(R.string.nyt_register_config_not_found), exc)
            }
        return null
    }

    private fun readConfigFile(): InputStream {
        return activity.resources.assets.open(CONFIG_FILE)
    }

    @Provides
    @ScopeActivity
    internal fun provideAlertDialogBuilder(activity: Activity): AlertDialog.Builder {
        return AlertDialog.Builder(activity)
    }

    @Provides
    @ScopeActivity
    internal fun provideRetrofit(client: OkHttpClient, @Named(GSON_RETROFIT) gson: Gson): GithubApi {
        return Retrofit.Builder()
                .client(client)
                .baseUrl(BuildConfig.GITHUB_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io()))
                .build()
                .create(GithubApi::class.java)
    }

}
